README
NAME

Devel::CallChecker - custom op checking attached to subroutines

DESCRIPTION

This module makes some new features of the Perl 5.14.0 C API available to
XS modules running on older versions of Perl.  The features are centred
around the function "cv_set_call_checker", which allows XS code to attach
a magical annotation to a Perl subroutine, resulting in resolvable calls
to that subroutine being mutated at compile time by arbitrary C code.
This module makes "cv_set_call_checker" and several supporting functions
available.  (It is possible to achieve the effect of "cv_set_call_checker"
from XS code on much earlier Perl versions, but it is painful to achieve
without the centralised facility.)

This module provides the implementation of the functions at runtime
(on Perls where they are not provided by the core), and also at compile
time supplies the C header file which provides access to the functions.

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2011, 2012, 2013 Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.006; 2013-09-21

  * bugfix: allow generated headers to work on API-compatible Perls 
    other than the specific version under which this module was installed

  * fix test for thread safety, which risked false negatives

  * avoid a C compiler warning in a test

version 0.005; 2012-02-11

  * be thread-safe, by idempotence and mutex control on op check hooking

  * load DynaLoader::Functions lazily, because it is only required at
    build time of users of this module, not required at all in normal
    runtime

  * avoid potential circular dependency chain, by requiring a version
    of DynaLoader::Functions that has reduced its dependencies

  * fix some C preprocessor directive indentation

version 0.004; 2012-02-01

  * in documentation, clarify that the header and linkable functions
    should be called at build time

  * add B::CallChecker to "see also" list

  * update tests to accept Perl 5.15.7's modified panic error messages

  * in Build.PL, declare incompatibility with pre-0.19
    B::Hooks::OP::Check, which doesn't play nicely around op check hooking

  * convert .cvsignore to .gitignore

version 0.003; 2011-05-29

  * bugfix: set up CV name links correctly for error messages from
    prototype checkers

  * bugfix: don't leak temporary GVs and CVs in prototype checkers

version 0.002; 2011-05-19

  * add callchecker_linkable constant to help users link with this module

  * fully document the C functions

  * avoid false test failures with parallel testing

  * correct abstract line

  * add Devel::CallParser to "see also" list

version 0.001; 2011-04-11

  * bugfix: use PERL_CALLCONV to achieve consistent ABI across compilers

  * port to Windows (GCC toolchain) and Cygwin, where additional linker
    magic is required to make importation from shared object work (MSVC
    and BCC on Windows presumably still don't generate the linkable
    version of the shared library)

version 0.000; 2011-04-03

  * initial released version


t/LoadXS.pm
package t::LoadXS;

use warnings;
use strict;

use DynaLoader ();
use ExtUtils::CBuilder ();
use ExtUtils::ParseXS ();
use File::Spec ();

sub loadable_lib_for_module($) {
	my($modname) = @_;
	# This logic is taken from DynaLoader.	It mixes native directory
	# names from @INC and Unix-style /-separated path syntax.
	# Does this really work everywhere?
	my @modparts = split(/::/,$modname);
	my $modend = $modparts[-1];
	my $modpname = join("/",@modparts);
	my $loadlib = DynaLoader::dl_findfile(
			(map { "-L$_/auto/$modpname" } @INC), $modend)
		or die "Can't locate loadable object ".
			"for module $modname in \@INC";
	return $loadlib;
}

my %linkablelib_finders = (
	MSWin32 => sub {
		if((my $basename = $_[0]) =~ s/\.[Dd][Ll][Ll]\z//) {
			foreach my $suffix (qw(.lib .a)) {
				my $impname = $basename.$suffix;
				return ($impname) if -e $impname;
			}
		}
		die "Can't locate linkable object for $_[0]";
	},
	cygwin => sub { ($_[0]) },
);

sub linkable_lib_for_module($) {
	my $finder = $linkablelib_finders{$^O} or return ();
	my($modname) = @_;
	return $finder->(loadable_lib_for_module($modname));
}

1;
our @todelete;
END { unlink @todelete; }

sub load_xs($$$) {
	my($basename, $dir, $importmods) = @_;
	my $xs_file = File::Spec->catdir("t", "$basename.xs");
	my $c_file = File::Spec->catdir("t", "$basename.c");
	ExtUtils::ParseXS::process_file(
		filename => $xs_file,
		output => $c_file,
	);
	push @todelete, $c_file;
	my $cb = ExtUtils::CBuilder->new(quiet => 1);
	my $o_file = $cb->compile(source => $c_file);
	push @todelete, $o_file;
	my @extra_libs = map { linkable_lib_for_module($_) } @$importmods;
	my($so_file, @so_tmps) = $cb->link(objects => [ $o_file, @extra_libs ],
						module_name => "t::$basename");
	push @todelete, $so_file, @so_tmps;
	my $boot_symbol = "boot_t__$basename";
	@DynaLoader::dl_require_symbols = ($boot_symbol);
	my $so_handle = DynaLoader::dl_load_file($so_file, 0);
	defined $so_handle or die(DynaLoader::dl_error());
	my $boot_func = DynaLoader::dl_find_symbol($so_handle, $boot_symbol);
	defined $boot_func or die "symbol $boot_symbol not found in $so_file";
	my $boot_perlname = "t::${basename}::bootstrap";
	DynaLoader::dl_install_xsub($boot_perlname, $boot_func, $so_file)->();
}

1;


t/WriteHeader.pm
package t::WriteHeader;

use warnings;
use strict;

use File::Spec ();
use IO::File 1.03 ();

our @todelete;
END { unlink @todelete; }

sub write_header($$) {
	my($basename, $outdir) = @_;
	require Devel::CallChecker;
	no strict "refs";
	my $content = &{"Devel::CallChecker::${basename}_h"}();
	my $h_file = File::Spec->catfile($outdir, "${basename}.h");
	push @todelete, $h_file;
	my $fh = IO::File->new($h_file, "w") or die $!;
	$fh->printflush($content) or die $!;
	$fh->close or die $!;
}

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4007
require 5.006;

    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";

      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');

      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }

      require Cwd;
      require File::Spec;
      require CPAN;

      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();

      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";

      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    use lib '_build/lib';
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    my $build_script = 'Build';
    $build_script .= '.com' if $^O eq 'VMS';
    exit(0) unless(-e $build_script); # cpantesters convention
    require MyModuleBuilder;
    Module::Build::Compat->write_makefile(build_class => 'MyModuleBuilder');


lib/Devel/CallChecker.pm
=head1 NAME

Devel::CallChecker - custom op checking attached to subroutines

=head1 SYNOPSIS

	# to generate header prior to XS compilation

	perl -MDevel::CallChecker=callchecker0_h \
		-e 'print callchecker0_h' > callchecker0.h

	# in Perl part of module

	use Devel::CallChecker;

	/* in XS */

	#include "callchecker0.h"

	cv_get_call_checker(cv, &ckfun, &ckobj);
	static OP *my_ckfun(pTHX_ OP *o, GV *namegv, SV *ckobj);
	cv_set_call_checker(cv, my_ckfun, ckobj);

=head1 DESCRIPTION

This module makes some new features of the Perl 5.14.0 C API available
to XS modules running on older versions of Perl.  The features are
centred around the function C<cv_set_call_checker>, which allows XS
code to attach a magical annotation to a Perl subroutine, resulting in
resolvable calls to that subroutine being mutated at compile time by
arbitrary C code.  This module makes C<cv_set_call_checker> and several
supporting functions available.  (It is possible to achieve the effect
of C<cv_set_call_checker> from XS code on much earlier Perl versions,
but it is painful to achieve without the centralised facility.)

This module provides the implementation of the functions at runtime (on
Perls where they are not provided by the core).  It also, at compile time,
supplies the C header file and link library which provide access to the
functions.  In normal use, L</callchecker0_h> and L</callchecker_linkable>
should be called at build time (not authoring time) for the module that
wishes to use the C functions.


=cut

package Devel::CallChecker;

{ use 5.006; }
use warnings;
use strict;

our $VERSION = "0.006";

use parent "Exporter";
our @EXPORT_OK = qw(callchecker0_h callchecker_linkable);

{
	require DynaLoader;
	local our @ISA = qw(DynaLoader);
	local *dl_load_flags = sub { 1 };
	__PACKAGE__->bootstrap($VERSION);
}

=head1 CONSTANTS

=over

=item callchecker0_h

Content of a C header file, intended to be named "C<callchecker0.h>".
It is to be included in XS code, and C<perl.h> must be included first.
When the XS module is loaded at runtime, the C<Devel::CallChecker>
module must be loaded first.  This will result in the Perl API functions
C<rv2cv_op_cv>, C<ck_entersub_args_list>, C<ck_entersub_args_proto>,
C<ck_entersub_args_proto_or_list>, C<cv_get_call_checker>, and
C<cv_set_call_checker>, as defined below and in the Perl 5.14.0 API,
being available to the XS code.

=item callchecker_linkable

List of names of files that must be used as additional objects when
linking an XS module that uses the C functions supplied by this module.
This list will be empty on many platforms.

=cut

sub callchecker_linkable() {
	require DynaLoader::Functions;
	DynaLoader::Functions->VERSION(0.001);
	return DynaLoader::Functions::linkable_for_module(__PACKAGE__);
}

=back

=head1 C FUNCTIONS

=over

=item rv2cv_op_cv

Examines an op, which is expected to identify a subroutine at runtime,
and attempts to determine at compile time which subroutine it identifies.
This is normally used during Perl compilation to determine whether
a prototype can be applied to a function call.  I<cvop> is the op
being considered, normally an C<rv2cv> op.  A pointer to the identified
subroutine is returned, if it could be determined statically, and a null
pointer is returned if it was not possible to determine statically.

Whether the subroutine is statically identifiable is determined in
accordance with the prevailing standards of the Perl version being used.
The same criteria are used that the core uses to determine whether to
apply a prototype to a subroutine call.  From version 5.11.2 onwards, the
subroutine can be determined if the RV that the C<rv2cv> is to operate
on is provided by a suitable C<gv> or C<const> op.  Prior to 5.11.2,
only a C<gv> op will do.  A C<gv> op is suitable if the GV's CV slot
is populated.  A C<const> op is suitable if the constant value must be
an RV pointing to a CV.  Details of this process may change in future
versions of Perl.

If the C<rv2cv> op has the C<OPpENTERSUB_AMPER> flag set then no attempt
is made to identify the subroutine statically: this flag is used to
suppress compile-time magic on a subroutine call, forcing it to use
default runtime behaviour.

If I<flags> has the bit C<RV2CVOPCV_MARK_EARLY> set, then the handling
of a GV reference is modified.  If a GV was examined and its CV slot was
found to be empty, then the C<gv> op has the C<OPpEARLY_CV> flag set.
If the op is not optimised away, and the CV slot is later populated with
a subroutine having a prototype, that flag eventually triggers the warning
"called too early to check prototype".

If I<flags> has the bit C<RV2CVOPCV_RETURN_NAME_GV> set, then instead
of returning a pointer to the subroutine it returns a pointer to the
GV giving the most appropriate name for the subroutine in this context.
Normally this is just the C<CvGV> of the subroutine, but for an anonymous
(C<CvANON>) subroutine that is referenced through a GV it will be the
referencing GV.  The resulting C<GV*> is cast to C<CV*> to be returned.
A null pointer is returned as usual if there is no statically-determinable
subroutine.

	CV *rv2cv_op_cv(OP *cvop, U32 flags)

=item cv_get_call_checker

Retrieves the function that will be used to fix up a call to I<cv>.
Specifically, the function is applied to an C<entersub> op tree for a
subroutine call, not marked with C<&>, where the callee can be identified
at compile time as I<cv>.

The C-level function pointer is returned in I<*ckfun_p>, and an SV
argument for it is returned in I<*ckobj_p>.  The function is intended
to be called in this manner:

    entersubop = (*ckfun_p)(aTHX_ entersubop, namegv, (*ckobj_p));

In this call, I<entersubop> is a pointer to the C<entersub> op,
which may be replaced by the check function, and I<namegv> is a GV
supplying the name that should be used by the check function to refer
to the callee of the C<entersub> op if it needs to emit any diagnostics.
It is permitted to apply the check function in non-standard situations,
such as to a call to a different subroutine or to a method call.

By default, the function is
L<Perl_ck_entersub_args_proto_or_list|/ck_entersub_args_proto_or_list>,
and the SV parameter is I<cv> itself.  This implements standard
prototype processing.  It can be changed, for a particular subroutine,
by L</cv_set_call_checker>.

	void cv_get_call_checker(CV *cv,
		Perl_call_checker *ckfun_p, SV **ckobj_p)

=item cv_set_call_checker

Sets the function that will be used to fix up a call to I<cv>.
Specifically, the function is applied to an C<entersub> op tree for a
subroutine call, not marked with C<&>, where the callee can be identified
at compile time as I<cv>.

The C-level function pointer is supplied in I<ckfun>, and an SV argument
for it is supplied in I<ckobj>.  The function is intended to be called
in this manner:

    entersubop = ckfun(aTHX_ entersubop, namegv, ckobj);

In this call, I<entersubop> is a pointer to the C<entersub> op,
which may be replaced by the check function, and I<namegv> is a GV
supplying the name that should be used by the check function to refer
to the callee of the C<entersub> op if it needs to emit any diagnostics.
It is permitted to apply the check function in non-standard situations,
such as to a call to a different subroutine or to a method call.

The current setting for a particular CV can be retrieved by
L</cv_get_call_checker>.

	void cv_set_call_checker(CV *cv, Perl_call_checker ckfun,
		SV *ckobj)

=item ck_entersub_args_list

Performs the default fixup of the arguments part of an C<entersub>
op tree.  This consists of applying list context to each of the
argument ops.  This is the standard treatment used on a call marked
with C<&>, or a method call, or a call through a subroutine reference,
or any other call where the callee can't be identified at compile time,
or a call where the callee has no prototype.

	OP *ck_entersub_args_list(OP *entersubop)

=item ck_entersub_args_proto

Performs the fixup of the arguments part of an C<entersub> op tree
based on a subroutine prototype.  This makes various modifications to
the argument ops, from applying context up to inserting C<refgen> ops,
and checking the number and syntactic types of arguments, as directed by
the prototype.  This is the standard treatment used on a subroutine call,
not marked with C<&>, where the callee can be identified at compile time
and has a prototype.

I<protosv> supplies the subroutine prototype to be applied to the call.
It may be a normal defined scalar, of which the string value will be used.
Alternatively, for convenience, it may be a subroutine object (a C<CV*>
that has been cast to C<SV*>) which has a prototype.  The prototype
supplied, in whichever form, does not need to match the actual callee
referenced by the op tree.

If the argument ops disagree with the prototype, for example by having
an unacceptable number of arguments, a valid op tree is returned anyway.
The error is reflected in the parser state, normally resulting in a single
exception at the top level of parsing which covers all the compilation
errors that occurred.  In the error message, the callee is referred to
by the name defined by the I<namegv> parameter.

	OP *ck_entersub_args_proto(OP *entersubop, GV *namegv,
		SV *protosv)

=item ck_entersub_args_proto_or_list

Performs the fixup of the arguments part of an C<entersub> op tree either
based on a subroutine prototype or using default list-context processing.
This is the standard treatment used on a subroutine call, not marked
with C<&>, where the callee can be identified at compile time.

I<protosv> supplies the subroutine prototype to be applied to the call,
or indicates that there is no prototype.  It may be a normal scalar,
in which case if it is defined then the string value will be used
as a prototype, and if it is undefined then there is no prototype.
Alternatively, for convenience, it may be a subroutine object (a C<CV*>
that has been cast to C<SV*>), of which the prototype will be used if it
has one.  The prototype (or lack thereof) supplied, in whichever form,
does not need to match the actual callee referenced by the op tree.

If the argument ops disagree with the prototype, for example by having
an unacceptable number of arguments, a valid op tree is returned anyway.
The error is reflected in the parser state, normally resulting in a single
exception at the top level of parsing which covers all the compilation
errors that occurred.  In the error message, the callee is referred to
by the name defined by the I<namegv> parameter.

	OP *ck_entersub_args_proto_or_list(OP *entersubop,
		GV *namegv, SV *protosv)

=back

=head1 SEE ALSO

L<B::CallChecker>,
L<Devel::CallParser>,
L<perlapi/cv_set_call_checker>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2011, 2012, 2013 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


Build.PL
{ use 5.006; }
use warnings;
use strict;

use Module::Build;

Module::Build->subclass(code => q{
	unless(__PACKAGE__->can("cbuilder")) {
		*cbuilder = sub { $_[0]->_cbuilder or die "no C support" };
	}
	sub link_c {
		no strict "refs";
		my($self, $spec) = @_;
		my $cb = $self->cbuilder;
		my $cbclass = ref($cb);
		my $orig_cb_prelink = $cb->can("prelink");
		local *{"${cbclass}::prelink"} = sub {
			use strict "refs";
			my($self, %args) = @_;
			if($args{dl_name} eq "Devel::CallChecker") {
				$args{dl_func_list} = [
					@{$args{dl_func_list}||[]},
					("$]" >= 5.013006 ? () : qw(
						xAd8NP3gxZglovQRL5Hn_roc0
						xAd8NP3gxZglovQRL5Hn_eal0
						xAd8NP3gxZglovQRL5Hn_eap0
						xAd8NP3gxZglovQRL5Hn_ean0
						xAd8NP3gxZglovQRL5Hn_gcc0
						xAd8NP3gxZglovQRL5Hn_scc0
					)),
				];
				$args{dl_funcs} ||= {};
				my $pname = $args{dl_name};
				unless(exists $args{dl_funcs}->{$pname}) {
					$args{dl_funcs} = {
						%{$args{dl_funcs}},
						$pname => [],
					};
				}
			}
			@_ = ($self, %args);
			goto &$orig_cb_prelink;
		};
		my($libfile, $impfile);
		if($^O eq "MSWin32") {
			my $dlext = $cb->{config}->{dlext};
			my $libext = $cb->{config}->{lib_ext};
			$libfile = $spec->{lib_file};
			($impfile = $libfile) =~ s/\.\Q$dlext\E\z/$libext/
				or die "can't generate import library name";
			unlink $libfile, $impfile
				unless $self->up_to_date($libfile, $impfile);
		}
		my $orig_cb_flk = $cb->can("format_linker_cmd");
		local *{"${cbclass}::format_linker_cmd"} = sub {
			use strict "refs";
			my($self, %spec) = @_;
			my @cmds = &$orig_cb_flk;
			my $cf = $self->{config};
			my $norm_libfile = $libfile;
			my $norm_impfile = $impfile;
			$self->normalize_filespecs(
				\$norm_libfile, \$norm_impfile);
			push @cmds, [
				$cf->{dlltool} || "dlltool",
				"--def", $spec{def_file},
				"--output-lib", $norm_impfile,
				"--dllname", $spec{basename}.".".$cf->{dlext},
				$spec{output},
			] if $spec{output} eq $norm_libfile;
			return @cmds;
		} if $cb->isa("ExtUtils::CBuilder::Platform::Windows::GCC");
		$self->SUPER::link_c($spec);
		if($^O eq "MSWin32") {
			die "failed to generate import library"
				unless -e $impfile;
			$self->add_to_cleanup($impfile);
		}
	}
})->new(
	module_name => "Devel::CallChecker",
	license => "perl",
	configure_requires => {
		"Module::Build" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	build_requires => {
		"DynaLoader" => 0,
		"ExtUtils::CBuilder" => "0.15",
		"ExtUtils::ParseXS" => 0,
		"File::Spec" => 0,
		"IO::File" => "1.03",
		"Module::Build" => 0,
		"Test::More" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	requires => {
		"DynaLoader" => 0,
		"DynaLoader::Functions" => 0,
		"Exporter" => 0,
		"parent" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	conflicts => {
		"B::Hooks::OP::Check" => "< 0.19",
	},
	dynamic_config => 0,
	meta_add => { distribution_type => "module" },
	create_makefile_pl => "passthrough",
	sign => 1,
)->create_build_script;

1;
